### PR TITLE
Remove consensus trait

### LINKED ISSUE #3198: Remove `Consensus` and (most of) `Beacon`
**Issue summary**

The `Consensus` trait should be removed. We currently only have one set of consensus rules (and if we want an additional set of rules in the future, we'll implement it differently.)

We have two implementations of `Beacon`: `DrandBeacon` and `MockBeacon`. Let's use a dynamic trait object instead of specifying the beacon at the type-level.

**Other information and links**

https://github.com/ChainSafe/forest/blob/4c7d560d016897af5a3ea69ae5defb31d974b9c4/src/chain_sync/consensus.rs#L32

https://github.com/ChainSafe/forest/blob/4c7d560d016897af5a3ea69ae5defb31d974b9c4/src/beacon/drand.rs#L138

### PR DESCRIPTION
**Pull request by creativcoder**
## Summary of changes

Changes introduced in this pull request:

- Removes consensus trait
- replaces Beacon constraint with dyn Beacon.

## Reference issue to close (if applicable)

Closes #3198 

## Change checklist

- [X] I have performed a self-review of my own code,
- [X] I have made corresponding changes to the documentation,
- [X] I have added tests that prove my fix is effective or that my feature works (if possible),
- [X] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

**Review by lemmih — CHANGES_REQUESTED** [review]
The discussion moved to Slack.

File: src/chain_sync/tipset_syncer.rs [hunk 1]
```diff
@@ -1307,15 +1300,16 @@ async fn validate_block<DB: Blockstore + Sync + Send + 'static, C: Consensus>(
     let v_block_store = state_manager.blockstore_owned();
     let v_base_tipset = Arc::clone(&base_tipset);
     let weight = header.weight().clone();
+    // let consensus = Arc::clone(&consensus);
```
Comment (lemmih) [inline src/chain_sync/tipset_syncer.rs:1303]:
```suggestion
```

File: src/chain_sync/tipset_syncer.rs [hunk 2]
```diff
@@ -1364,6 +1358,7 @@ async fn validate_block<DB: Blockstore + Sync + Send + 'static, C: Consensus>(
     }));
 
     let v_block = block.clone();
+    // let consensus = consensus.clone();
```
Comment (lemmih) [inline src/chain_sync/tipset_syncer.rs:1361]:
```suggestion
```

File: src/chain_sync/tipset_syncer.rs [hunk 3]
```diff
@@ -1213,11 +1206,11 @@ async fn validate_tipset<DB: Blockstore + Send + Sync + 'static, C: Consensus>(
 /// * Checking that the messages in the block correspond to the agreed upon
 ///   total ordering
 /// * That the block is a deterministic derivative of the underlying consensus
-async fn validate_block<DB: Blockstore + Sync + Send + 'static, C: Consensus>(
-    consensus: Arc<C>,
+async fn validate_block<DB: Blockstore + Sync + Send + 'static>(
+    consensus: Arc<FilecoinConsensus>,
```
Comment (lemmih) [inline src/chain_sync/tipset_syncer.rs:1210]: The `consensus` argument is not required here. You can create a new `FilecoinConsensus` structure from the beacon in the state manager. Removing the `consensus` argument will have a ripple effect on the rest of the code.

**Review by lemmih — CHANGES_REQUESTED** [review]

File: src/cli/subcommands/snapshot_cmd.rs [hunk 4]
```diff
@@ -453,7 +453,7 @@ where
     load_actor_bundles(&db).await?;
 
     // Set proof parameter data dir and make sure the proofs are available
-    if cns::FETCH_PARAMS {
+    if FETCH_PARAMS {
```
Comment (lemmih) [inline src/cli/subcommands/snapshot_cmd.rs:456]: `FETCH_PARAMS` is always true. Let's unroll the code.

File: src/chain_sync/tipset_syncer.rs [hunk 5]
```diff
@@ -1097,16 +1075,15 @@ async fn sync_messages_check_state<DB: Blockstore + Send + Sync + 'static, C: Co
         // Chunk tipsets in batches (default batch size is 8)
         .chunks(request_window)
         // Request batches from the p2p network
-        .map(|batch| fetch_batch::<_, C>(batch, &network, db))
+        .map(|batch| fetch_batch::<_>(batch, &network, db))
```
Comment (lemmih) [inline src/chain_sync/tipset_syncer.rs:1078]:
```suggestion
        .map(|batch| fetch_batch(batch, &network, db))
```

File: src/chain_sync/tipset_syncer.rs [hunk 6]
```diff
@@ -1097,16 +1075,15 @@ async fn sync_messages_check_state<DB: Blockstore + Send + Sync + 'static, C: Co
...
         // run 64 batches concurrently
         .buffered(64)
         // validate each full tipset in each batch
         .try_for_each(|batch| async {
             for full_tipset in batch {
                 let current_epoch = full_tipset.epoch();
                 let timer = metrics::TIPSET_PROCESSING_TIME.start_timer();
-                validate_tipset::<_, C>(
-                    consensus.clone(),
+                validate_tipset::<_>(
```
Comment (lemmih) [inline src/chain_sync/tipset_syncer.rs:1086]:
```suggestion
                validate_tipset(
```

File: src/chain_sync/tipset_syncer.rs [hunk 7]
```diff
@@ -1156,11 +1132,8 @@ async fn validate_tipset<DB: Blockstore + Send + Sync + 'static, C: Consensus>(
     debug!("Tipset keys: {:?}", full_tipset_key.cids);
 
     for b in blocks {
-        let validation_fn = tokio::task::spawn(validate_block::<_, C>(
-            consensus.clone(),
-            state_manager.clone(),
-            Arc::new(b),
-        ));
+        let validation_fn =
+            tokio::task::spawn(validate_block::<_>(state_manager.clone(), Arc::new(b)));
```
Comment (lemmih) [inline src/chain_sync/tipset_syncer.rs:1136]:
```suggestion
            tokio::task::spawn(validate_block(state_manager.clone(), Arc::new(b)));
```

File: src/chain_sync/tipset_syncer.rs [hunk 8]
```diff
@@ -1275,7 +1248,7 @@ async fn validate_block<DB: Blockstore + Sync + Send + 'static, C: Consensus>(
     let validations = FuturesUnordered::new();
 
     // Check block messages
-    validations.push(tokio::task::spawn(check_block_messages::<_, C>(
+    validations.push(tokio::task::spawn(check_block_messages::<_>(
```
Comment (lemmih) [inline src/chain_sync/tipset_syncer.rs:1251]:
```suggestion
    validations.push(tokio::task::spawn(check_block_messages(
```

**Review by lemmih — APPROVED** [review]

**Review by sudo-shashank — APPROVED** [review]

**Pull request merged**